sdk/python/python.go

```diff
@@ -20,19 +20,42 @@ import (
 	"io"
 	"os"
 	"os/exec"
+	"path"
 	"path/filepath"
 	"runtime"
 	"strings"
 
 	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/v3/util/yamlutil"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"gopkg.in/yaml.v3"
 )
 
 const (
 	windows             = "windows"
 	pythonShimCmdFormat = "pulumi-%s-shim.cmd"
 )
 
+type Toolchain string 
+const (
+	Poetry Toolchain       = "poetry"
+	Venv Toolchain         = "venv"
+	Pyenv Toolchain        = "pyenv"
+)
+
+func ParseToolchain(toolchain string) (Toolchain, error) {
+	switch toolchain {
+	case "poetry":
+		return Poetry, nil
+	case "venv":
+		return Venv, nil
+	case "pyenv":
+		return Pyenv, nil
+	default:
+		return "", fmt.Errorf("unknown toolchain %q", toolchain)
+	}
+}
```

Comment:
Okay, I'm aware this function _looks_ ridiculous.  However, toolchain crosses an I/O boundary, and I'd generally prefer to err on the side of safety on both sides of that boundary.